src/databricks/labs/ucx/azure/azure_credentials.py

```diff
@@ -0,0 +1,283 @@
+        # fetch sp client_secret if any
+        sp_list_with_secret = self._fetch_client_secret(filtered_sp_list)
+        self._final_sp_list = sp_list_with_secret
+        # output the action plan for customer to confirm
+        self._print_action_plan(sp_list_with_secret)
+
+    def _create_storage_credential(self, sp_migration: ServicePrincipalMigrationInfo):
+        # prepare the storage credential properties
+        name = sp_migration.service_principal.principal
+        azure_service_principal = AzureServicePrincipal(
+            directory_id=sp_migration.service_principal.directory_id,
+            application_id=sp_migration.service_principal.client_id,
+            client_secret=sp_migration.client_secret,
+        )
+        comment = f"Created by UCX during migration to UC using Azure Service Principal: {sp_migration.service_principal.principal}"
+        read_only = False
+        if sp_migration.service_principal.privilege == Privilege.READ_FILES.value:
+            read_only = True
+        # create the storage credential
+        storage_credential = self._ws.storage_credentials.create(
+            name=name, azure_service_principal=azure_service_principal, comment=comment, read_only=read_only
+        )
+
+        validation_result = self._validate_storage_credential(
+            storage_credential, sp_migration.service_principal.prefix, read_only
+        )
+        return validation_result
+
+    def _validate_storage_credential(
+        self, storage_credential, location: str, read_only: bool
+    ) -> StorageCredentialValidationResult:
+        # storage_credential validation creates a temp UC external location, which cannot overlap with
+        # existing UC external locations. So add a sub folder to the validation location just in case
+        try:
+            validation = self._ws.storage_credentials.validate(
+                storage_credential_name=storage_credential.name, url=location, read_only=read_only
+            )
+            return StorageCredentialValidationResult.from_storage_credential_validation(storage_credential, validation)
+        except InvalidParameterValue:
+            logger.warning(
+                "There is an existing external location overlaps with the prefix that is mapped to the service principal and used for validating the migrated storage credential. Skip the validation"
+            )
+            return StorageCredentialValidationResult.from_storage_credential_validation(
+                storage_credential,
+                ValidateStorageCredentialResponse(
+                    is_dir=None,
+                    results=[
+                        ValidationResult(
+                            message="The validation is skipped because an existing external location overlaps with the location used for validation."
+                        )
+                    ],
+                ),
+            )
+
+    def execute_migration(self, prompts: Prompts):
```

Comment:
```suggestion
    def apply(self, prompts: Prompts):
```
Better name:) or run()

Reply:
changed to run() :)